### PR TITLE
Factorial calculator

### PR DESCRIPTION
**Pull request by linahKhayri**
---
name: solution and tests for factorial_calculator function
about: A template PR for code review with a checklist
---

This challenge is to create a function that calculates the factorial of a non-negative integer. The function should

Return the factorial of the given number n, which is defined as n! = n * (n-1) * ... * 1.
Handle the base cases where 0! = 1 and 1! = 1.
Raise an assertion error if the input is not an integer or is a negative number, as factorials are only defined for positive integers.
Handle edge cases where the input is a float that represents a whole number, or the input is a large number.

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called in the function file

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/tests/test_factorial_calculator.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+        with self.assertRaises(AssertionError):
+            factorial_calculator(-5)
+
+    def test_fractional_float_factorial_calculator(self):
+        """should raise an error if the input num is not a whole float"""
+        with self.assertRaises(AssertionError):
+            factorial_calculator(3.5)
+
+    def test_non_numeric_factorial_calculator(self):
+        """should raise an error if the input num is a negative number"""
```
Comment (Mohamed-Elnageeb) [inline solutions/tests/test_factorial_calculator.py:92]: I think  you meant "....input is not a number" 😅

**Review by Mohamed-Elnageeb — COMMENTED** [review]
good job linah, your solution is complete and well formated, I like how you used recursion to solve the problem. The files check all the points on evan's check list 👌. 
only a small comment I have highlighted.

**Comment by linahKhayri** [comment]
> good job linah, your solution is complete and well formated, I like how you used recursion to solve the problem. The files check all the points on evan's check list 👌. only a small comment I have highlighted.

Thanks a lot for the amazing feedback! 
I'm glad I'm on the right track. Appreciate the note and I'll check it out now.

File: solutions/tests/test_factorial_calculator.py (hunk 1 above)
Comment (linahKhayri) [reply solutions/tests/test_factorial_calculator.py:92]: oh yes, my bad 😅
Thankss again Mohamed, will fix it right away

File: solutions/factorial_calculator.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    """
+    This function calculates the factorial of a non-negative integer or whole float using recursion.
+
+    The factorial of a non-negative integer n is the product of all positive integers
+    less than or equal to num until we reach one.
+    For example, the factorial of num! = num * (num-1) * ... * 1.
+    Also 0 and 1 are special cases when it comes to calculating the factorial
+    as their factorial is 1
+
+    Parameters:
```
Comment (robiel0143) [inline solutions/factorial_calculator.py:26]: this is to be applied to all ,  indentation on the comments like
Parameters:
          num:........

File: solutions/factorial_calculator.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    num : int or float
+    A non-negative integer or whole float whose factorial is to be computed.
+    Whole floats are floats without a fractional part, such as 1.0 or 5.0.
+
+    Returns:
```
Comment (robiel0143) [inline solutions/factorial_calculator.py:31]: Returns -> (notation) its okay but the notation would be good.

**Review by robiel0143 — COMMENTED** [review]
Well prepared, there are some comments that you have to arrange.

File: solutions/factorial_calculator.py (hunk 3 above)
Comment (linahKhayri) [reply solutions/factorial_calculator.py:31]: Thank you Robel 🙌
but I think in docstring it's better to refer to the return type in words , as the -> notation is used in function definition

**Comment by linahKhayri** [comment]
> Well prepared, there are some comments that you have to arrange.

Thankss for the feedback 💖

File: solutions/factorial_calculator.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
+#!/usr/bin/env python3
```
Comment (ViktoriyaHaiduk) [inline solutions/factorial_calculator.py:1]: @linahKhayri Great work! Well structured and clear code. I like number of unit tests.

**Review by ViktoriyaHaiduk — APPROVED** [review]
Great work! Well structured and clear code. I like number of unit tests.

**Comment by linahKhayri** [comment]
> Great work! Well structured and clear code. I like number of unit tests.

Thankss @ViktoriyaHaiduk! appreciate your feedback 💖

**Review by Mohamed-Elnageeb — COMMENTED** [review]
Well done linah!
Your solution checks all the boxes.

**Review by Mohamed-Elnageeb — APPROVED** [review]

File: solutions/factorial_calculator.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    # Base case for recursion: factorial of 0 or 1 is 1
+    if num == 0:  # Base case 1
+        return 1  # turn-around 1
+
+    if num == 1:  # Base case 2
+        return 1  # turn-around 2
+
+    # Recursive case: factorial(num) = num * factorial(num-1)
+    # breaking-down num | Build-up by multiplying
+    return num * factorial_calculator(num - 1)
```
Comment (Mohamed-Elnageeb) [inline solutions/factorial_calculator.py:69]: Hi , sorry for late review but i just realised Evan added another point on the checklist . So the thing is we shouldn't call the actual function inside the file itself.I believe we should define another helper functiom for recursion.I asked him and am waiting for his response

**Review by Mohamed-Elnageeb — COMMENTED** [review]
There is one checklist point that needs to  be reviewed
